src/controller/ICDRegistrationDelegate.h

```diff
@@ -0,0 +1,75 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <cinttypes>
+
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace Controller {
+
+/**
+ * @brief
+ * ICDRegistrationDelegate is an interface that provides a callback method for generating symmetric keys and reporting ICD
+ * registration readiness.
+ *
+ * This delegate is used by the commissioner to generate symmetric keys and complete ICD registration with the ICD
+ * device and to report when the device is ready for subscription.
+ */
+class ICDRegistrationDelegate
+{
+public:
+    static constexpr size_t kKeySize = Crypto::kAES_CCM128_Key_Length;
+    using ICDKey                     = FixedByteSpan<kKeySize>;
+
+    /// The type of callback function that is called when symmetric key generation is completed.
+    typedef void (*OnSymmetricKeyGenerationCompleted)(void * context, CHIP_ERROR error, NodeId checkInNodeId, uint64_t subjectId,
+                                                      ICDKey key);
+
+    virtual ~ICDRegistrationDelegate() = default;
+
+    /**
+     * @brief
+     *   Called when the operational commissioning completed, the controller may register it with the ICD device.
+     * The delegate can generate the symmetric key and call commissioner->OnSymmetricKeyGenerated to complete ICD
+     * registration.
+     *   This delegate supports Commissioner Self-Registration and External Controller Registration flow.
+     *
+     * @param[in] device       The node id of the device.
+     * @param[in] onCompletion The callback object to be called when symmetric key generation is completed. The
+     *                         `onCompletion` is expected to be valid through the key generation process. Implementations
+     *                         should drop the reference of `onCompletion` after it is called.
```

Comment:
I don't think this is correct.  If someone aborts commissioning on the controller, or shuts it down and destroys it, none of that will be true.

This needs to document the _actual_ safety guarantees here.  Just holding the raw onCompletion pointer is not safe.

Reply:
I realized that I'm making it over-complex in this PR.

We can just let the controller to pregenerate symmetric key, and commit it when we completes the registration (or remove the record if the deivce does not need registration).

And OnICDRegistrationComplete can be moved to DevicePairingDelegate with a empty implementation.